airbyte-integrations/connectors/destination-s3/src/main/java/io/airbyte/integrations/destination/s3/parquet/JsonSchemaConverter.java

```diff
@@ -0,0 +1,174 @@
+import java.util.stream.Collectors;
+import javax.annotation.Nullable;
+import org.apache.avro.LogicalTypes;
+import org.apache.avro.Schema;
+import org.apache.avro.Schema.Type;
+import org.apache.avro.SchemaBuilder;
+import org.apache.avro.SchemaBuilder.RecordBuilder;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The main function of this class is to convert a JsonSchema to Avro schema. It can also
+ * standardize schema names, and keep track of a mapping from the original names to the
+ * standardized ones.
+ */
+public class JsonSchemaConverter {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(JsonSchemaConverter.class);
+  public static final Schema UUID_SCHEMA = LogicalTypes.uuid()
+      .addToSchema(Schema.create(Type.STRING));
+  private static final Schema TIMESTAMP_MILLIS_SCHEMA = LogicalTypes.timestampMillis()
+      .addToSchema(Schema.create(Type.LONG));
+  private static final StandardNameTransformer NAME_TRANSFORMER = new StandardNameTransformer();
+
+  private final Map<String, String> standardizedNames = new HashMap<>();
+
+  public Map<String, String> getStandardizedNames() {
+    return standardizedNames;
+  }
+
+  /**
+   * @return - Avro schema based on the input {@code jsonSchema}.
+   */
+  public Schema getAvroSchema(JsonNode jsonSchema,
+                              String name,
+                              @Nullable String namespace,
+                              boolean appendAirbyteFields) {
+    String stdName = NAME_TRANSFORMER.getIdentifier(name);
+    RecordBuilder<Schema> builder = SchemaBuilder.record(stdName);
+    if (!stdName.equals(name)) {
+      standardizedNames.put(name, stdName);
+      LOGGER.warn("Schema name contains illegal character(s) and is standardized: {} -> {}", name,
+          stdName);
+      builder = builder.doc(
+          String.format("%s%s%s",
+              S3ParquetConstants.DOC_KEY_ORIGINAL_NAME,
+              S3ParquetConstants.DOC_KEY_VALUE_DELIMITER,
+              name));
+    }
+    if (namespace != null) {
+      builder = builder.namespace(namespace);
+    }
+
+    JsonNode properties = jsonSchema.get("properties");
+    List<String> fieldNames = new ArrayList<>(MoreIterators.toList(properties.fieldNames()));
+
+    SchemaBuilder.FieldAssembler<Schema> assembler = builder.fields();
+
+    if (appendAirbyteFields) {
+      assembler = assembler.name(JavaBaseConstants.COLUMN_NAME_AB_ID).type(UUID_SCHEMA).noDefault();
+      assembler = assembler.name(JavaBaseConstants.COLUMN_NAME_EMITTED_AT)
+          .type(TIMESTAMP_MILLIS_SCHEMA).noDefault();
+    }
+
+    for (String fieldName : fieldNames) {
+      String stdFieldName = NAME_TRANSFORMER.getIdentifier(fieldName);
+      JsonNode fieldDefinition = properties.get(fieldName);
+      SchemaBuilder.FieldBuilder<Schema> fieldBuilder = assembler.name(stdFieldName);
+      if (!stdFieldName.equals(fieldName)) {
+        standardizedNames.put(fieldName, stdFieldName);
```

Comment:
is there a concern of collision between field names and stream names? I suppose since we're using the same transformer it shouldn't be an issue (even if there is a collision the mapped name is the same) but it should maybe be separate maps for future proofing

Reply:
The stream name will be come the root level record name, and I just tested that record name and field name can be the same.

You are right that the standardized names will be the same any way, even if there is a collision. So I don't think separating the maps is necessary.